### DISCUSSION TITLE
Converting string to Resource, IRI, Value

### DISCUSSION THREAD

**Discussion by D063520** [eclipse-rdf4j/rdf4j#3547] [Q&A]
Hi,

I have the following problem. I have a string that I want to parse into a Resource, IRI or Value. I know about the [ValueFactory](https://rdf4j.org/javadoc/latest/index.html?org/eclipse/rdf4j/model/ValueFactory.html) that has methods to create IRIs, Literals and so on. The problem is that I do not know if the string represents a blank node, literal or URI.

The task is very similar to parsing a file, besides that here it is just one "node".

Does anyone has an idea. The only one I had was to do:
if starts with "_" -> value factory.createBlank
if starts with "\"" -> value factory.createLiteral

and so on ....

Thank you
D063520

**Comment by barthanssens** [comment]
Hi,

hmm well it actually depends on the format, because different formats may write blank nodes and values differently (eg "_:" for blank node and '<' to indicate the start of an URI in N-Triples, while that's not the case in e.g. JSON-LD)

Could you provide some additional background where the string you want to parse comes from (is it something a user has to provide in a web interface, or is it read from a file or something ?)

Best regards

Bart

**Reply by D063520** [reply]
In another part of my code I really have to parse back a string literal to a literal object. Basically I have:

- an implementation of a org.eclipse.rdf4j.model.impl.SimpleLiteral object, from which I call [toString() ](https://github.com/eclipse/rdf4j/blob/75b97f6106e6a73776d7f0f663f18a3edfb0a4ae/core/model/src/main/java/org/eclipse/rdf4j/model/impl/SimpleLiteral.java#L216)

- some logic that stores the string in a separate data-structure  

- parse the strings back. What I currently do is the following

```
    public static Literal parseLiteral(String nTriplesLiteral, ValueFactory valueFactory) throws IllegalArgumentException {
        if (nTriplesLiteral.startsWith("\"")) {
            int endLabelIdx = findEndOfLabel(nTriplesLiteral);
            if (endLabelIdx != -1) {
                int startLangIdx = nTriplesLiteral.indexOf(64, endLabelIdx);
                int startDtIdx = nTriplesLiteral.indexOf("^^", endLabelIdx);
                if (startLangIdx != -1 && startDtIdx != -1) {
                    throw new IllegalArgumentException("Literals can not have both a language and a datatype");
                }

                String label = nTriplesLiteral.substring(1, endLabelIdx);
                //label = unescapeString(label);
                String datatype;
                if (startLangIdx != -1) {
                    datatype = nTriplesLiteral.substring(startLangIdx + 1);
                    return valueFactory.createLiteral(label, datatype);
                }

                if (startDtIdx != -1) {
                    datatype = nTriplesLiteral.substring(startDtIdx + 2);
                    IRI dtURI = parseURI(datatype, valueFactory);
                    return valueFactory.createLiteral(label, dtURI);
                }

                return valueFactory.createLiteral(label);
            }
        }
        throw new IllegalArgumentException("Not a legal N-Triples literal: " + nTriplesLiteral);
    }

    private static int findEndOfLabel(String nTriplesLiteral) {
        for (int i = nTriplesLiteral.length() - 1; i >= 0; --i) {
            char c = nTriplesLiteral.charAt(i);
            if (c == '"') {
                return i;
            }
        }
        return -1;
    }
```

Do you see a more elegant solution? My assumption is that what comes out of the toString() is an ntriples formatted string.

**Reply by abrokenjester** [reply]
Instead of implementing all this yourself, you could also just use `org.eclipse.rdf4j.rio.helpers.NTriplesUtil`, which has various ready-made parse methods for processing NTriples strings into Literals, IRIs, or blank nodes. 

As for relying on `Literal#toString()` - it happens to be the case that it returns an N-Triples string in most implementations, but _this is not guaranteed_ - `toString()` is really only intended to be used for human-readable representation. If you want to ensure you get a N-Triples serialization, use `NTriplesUtil.toNTriplesString` instead.

**Comment by D063520** [comment]
Hi!

Thank you for your help! Basically the strings are coming from objects of the type Resource, IRI, Value. I discovered that I can check what they are by asking "instanceof BNode" and so on. So I found a solution to my problem.

Proposal. Why do we do not add some methods like "isBlankNode()" to the different classes to check this? At least the complier will propose it and I think people will find the information faster?

Thank you again
D063520

**Reply by barthanssens** [reply]
Ah, well, the Value interface already has `isBNode`, `isIRI`, `isLiteral` methods (and IRI/Literal/Resource/BNode... are subinterfaces, so they inherit these methods).

https://rdf4j.org/javadoc/latest/index.html?org/eclipse/rdf4j/model/Value.html

Is it possible to simply avoid converting the values to string ?

**Reply by D063520** [reply]
hallo, you are right! I did not see it. Thank you again for the pointer! About my problem, I think it is not possible ... I really need strings ....
